contracts/near/src/lib.rs

```diff
@@ -0,0 +1,489 @@
+        // deserialize into an Action.
+        let vaa = hex::decode(&vaa).unwrap();
+        let (_, rest): (structs::Vaa<()>, _) = vaa_payload::from_slice_with_payload(&vaa).unwrap();
+
+        // Attempt to deserialize the Batch of Price Attestations.
+        let bytes = &mut Cursor::new(rest);
+        let batch = BatchPriceAttestation::deserialize(bytes).unwrap();
+
+        // Verify the PriceAttestation's are new enough, and if so, store them.
+        let mut count_updates = 0;
+        for price_attestation in &batch.price_attestations {
+            if self.update_price_feed_if_new(PriceFeed::from(price_attestation)) {
+                count_updates += 1;
+            }
+        }
+
+        log!(
+            r#"
+            {{
+                "standard": "pyth",
+                "version":  "1.0",
+                "event":    "BatchAttest",
+                "data":     {{
+                    "count": {}
+                }}
+            }}
+        "#,
+            count_updates
+        );
+
+        // Refund storage difference to `account_id` after storage execution.
+        self.refund_storage_usage(
+            account_id,
+            storage,
+            env::storage_usage(),
+            env::attached_deposit(),
+        )
+    }
+
+    /// Process a Batch of Price Attestations.
+    ///
+    /// This receives price updates, verifies them, and stores the attestations in the contract
+    /// state allowing other contracts to query them.
+    #[private]
+    pub fn attest_prices(&mut self, batch: Vec<u8>) {
+        // Attempt to deserialize the Batch of Price Attestations.
+        let bytes = &mut Cursor::new(batch);
+        let batch = BatchPriceAttestation::deserialize(bytes).unwrap();
+
+        // Verify the PriceAttestation's are new enough, and if so, store them.
+        let mut count_updates = 0;
+        for price_attestation in &batch.price_attestations {
+            if self.update_price_feed_if_new(PriceFeed::from(price_attestation)) {
+                count_updates += 1;
+            }
+        }
+
+        log!(
+            r#"
+            {{
+                "standard": "pyth",
+                "version":  "1.0",
+                "event":    "BatchAttest",
+                "data":     {{
+                    "count": {}
+                }}
+            }}
+        "#,
+            count_updates
+        );
+    }
+
+    /// Read the list of accepted `Source` chains for a price attestation.
+    pub fn get_sources(&self) -> Vec<Source> {
+        self.sources.iter().collect()
+    }
+
+    /// Read the current Price for a given Price Feed.
+    ///
+    /// if the price is past the staleness threshold the trading status is set to Unknown before
+    /// returning to the caller.
+    pub fn get_price(&self, price_id: Identifier) -> Option<PriceFeed> {
+        self.prices.get(&price_id).and_then(|mut price| {
```

Comment:
The API for reading prices should be equivalent to the other contracts. Instead of the notion of "price status", we simply return `None` here (or maybe an informative `Err`) if a recent-enough price is not available. We then provide other functions to use a different recency threshold if appropiate for them. This is safer, because it forces the user to consider recency.

It would be good if we could have feature parity with https://github.com/pyth-network/pyth-crosschain/blob/main/aptos/contracts/sources/pyth.move:
- price_feed_exists
- get_price (with default staleness threshold)
- get_price_no_older_than (with custom staleness threshold)
- get_price_unsafe (return price regardless of age)